pymc3/distributions/mixture.py

```diff
@@ -12,30 +12,75 @@
 #   See the License for the specific language governing permissions and
 #   limitations under the License.
 
-from collections.abc import Iterable
-
-import aesara
 import aesara.tensor as at
 import numpy as np
 
-from pymc3.aesaraf import _conversion_map, take_along_axis
+from aesara.tensor.random.op import RandomVariable
+
+from pymc3.aesaraf import take_along_axis
+from pymc3.distributions import _logp
 from pymc3.distributions.continuous import Normal, get_tau_sigma
-from pymc3.distributions.dist_math import bound
-from pymc3.distributions.distribution import Discrete, Distribution
+from pymc3.distributions.dist_math import bound, random_choice
+from pymc3.distributions.distribution import Distribution
 from pymc3.distributions.shape_utils import to_tuple
 from pymc3.math import logsumexp
 
 __all__ = ["Mixture", "NormalMixture", "MixtureSameFamily"]
 
 
-def all_discrete(comp_dists):
-    """
-    Determine if all distributions in comp_dists are discrete
-    """
-    if isinstance(comp_dists, Distribution):
-        return isinstance(comp_dists, Discrete)
-    else:
-        return all(isinstance(comp_dist, Discrete) for comp_dist in comp_dists)
+class MixtureRV(RandomVariable):
+    def __init__(self, w=None, comp_dist=None):
+        self.name = "mixture"
+        self.comp_dist = comp_dist
+        self.w = w
+        self.ndim_supp = 0
+        # Needs to be a tuple fails hashing when a list. Needs investigation
+        self.ndims_params = (0,)
+        self.dtype = "floatX"
+        self.inplace = False
+        self._print_name = ("Mixture", "\\operatorname{Mixture}")
+
+    def rng_fn(self, rng, *args):
+        size = args[-1]
+        # The last dimension is of number of componenets so we trim it off
+        size = size[:-1] if size else None
+        orig_size = size
+        # Flatten the size to a one dimensional one
+        size = np.prod(size)
+
+        comp_dist = self.comp_dist
+        w = self.w
+
+        if isinstance(w, (list, tuple)):
+            w = np.array(w)
+        # Take care of case when w is a distribution
+
+        _size = size if size else 1
+        w_samples = random_choice(p=w, size=_size)
+
+        # Draw samples from each component
+        _samples = []
+        for dist in comp_dist:
+            curr_op = dist.owner.op
+            curr_inputs = dist.owner.inputs
+
+            if size is not None:
+                curr_inputs[1] = at.as_tensor_variable(size)
+
+            curr_op_random = at.as_tensor_variable(curr_op.make_node(*curr_inputs))
+            curr_op_random = curr_op_random.eval()
```

Comment:
I am not very optimistic about this approach, but in any case, you can try to call the `dist.rng_fn` directly to avoid creating and evaluating symbolic arguments in here.